File: grpc-java/grpc-java-app/src/test/java/com/example/ServerClientReactorTest.java
package com.example;

import com.example.foo.bar.HelloReply;
import com.example.foo.bar.HelloRequest;
import com.example.foo.bar.ReactorGreeter1Grpc;
import com.example.interceptors.Headers;
import com.example.services.ClientService;
import com.example.services.ServerService;
import com.example.utils.GrpcUtils;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Mono;

import static com.example.utils.LoggingUtils.log;
import static com.example.utils.TestUtils.MSG1;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;

class ServerClientReactorTest {
    @Test
    void test() throws Exception {
        var serverService = mock(ServerService.class);
        var clientService = mock(ClientService.class);

        try (var ignored = GrpcUtils.startServer(new ServerReactorImpl(serverService))) {
            try (var channel = GrpcUtils.createChannel()) {
                var stub = GrpcUtils.createStub(channel.channel(), ReactorGreeter1Grpc::newReactorStub);
                stub.sayHello(HelloRequest.newBuilder().setMsg(MSG1).build())
                    .doOnNext(reply -> {
                        log("newReactorStub.sayHello %s", reply);
                        clientService.call(reply.getMsg());
                    }).block();
            }
        }

        verify(serverService).call(MSG1);

        verify(clientService).call(MSG1.repeat(3));

        verifyNoMoreInteractions(serverService, clientService);
    }

    private static class ServerReactorImpl extends ReactorGreeter1Grpc.Greeter1ImplBase {
        private final ServerService service;

        public ServerReactorImpl(ServerService service) {
            this.service = service;
        }

        @Override
        public Mono<HelloReply> sayHello(HelloRequest request) {
            return Mono.fromRunnable(() -> {
                String requestId = Headers.REQUEST_ID_CTX_KEY.get();
                log("[%s] ServerReactorImpl.sayHello %s", requestId, request);
                service.call(request.getMsg());
            }).then(Mono.defer(() -> {
                var reply = HelloReply.newBuilder().setMsg(request.getMsg().repeat(3)).build();
                return Mono.just(reply);
            }));
        }

        @Override
        public Mono<HelloReply> sayHello(Mono<HelloRequest> request) {
            return request.flatMap(this::sayHello);
        }
    }
}

File: demo-grpc-proto/src/test/kotlin/com/example/greeter1/ServerClientReactorTest.kt
package com.example.greeter1

import com.example.util.*
import foo.bar.Greeter
import foo.bar.ReactorGreeter1Grpc
import io.grpc.ClientInterceptors
import io.grpc.Grpc
import io.grpc.ServerInterceptors
import org.junit.jupiter.api.Test
import org.mockito.kotlin.mock
import org.mockito.kotlin.timeout
import org.mockito.kotlin.verify
import org.mockito.kotlin.verifyNoMoreInteractions
import reactor.core.publisher.Mono

private val PORT = findFreePort()

class ServerClientReactorTest {
    @Test
    fun test() {
        val serverService = mock<ServerService>()
        val server = Grpc.newServerBuilderForPort(PORT, serverCredentials(false))
            .addService(ServerInterceptors.intercept(ServerReactorImpl(serverService), HeaderServerInterceptor()))
            .build()
            .start()
        println("Server started on port: $PORT")

        val clientService = mock<ClientService>()
        val channel = Grpc.newChannelBuilder("localhost:$PORT", clientCredentials(false)).build()
        val stub = ReactorGreeter1Grpc.newReactorStub(ClientInterceptors.intercept(channel, HeaderClientInterceptor()))
        stub.sayHello(Greeter.HelloRequest.newBuilder().setMsg(MSG1).build())
            .doOnNext { reply ->
                println("newReactorStub.sayHello $reply".trim() + " thread: ${Thread.currentThread()}")
                clientService.call(reply.msg)
            }
            .subscribe()

        verify(serverService, timeout(5000)).call(MSG1)

        verify(clientService, timeout(5000)).call(MSG1.repeat(3))

        verifyNoMoreInteractions(serverService, clientService)

        server.shutdown()
        server.awaitTermination()
    }
}

private class ServerReactorImpl(private val service: ServerService) : ReactorGreeter1Grpc.Greeter1ImplBase() {
    override fun sayHello(request: Greeter.HelloRequest): Mono<Greeter.HelloReply> {
        return Mono.fromRunnable<Greeter.HelloReply> {
            println("ServerReactorImpl.sayHello $request".trim() + " thread: ${Thread.currentThread()}")
            service.call(request.msg)
        }.then(Mono.defer {
            val reply = Greeter.HelloReply.newBuilder().setMsg(request.msg.repeat(3)).build()
            Mono.just(reply)
        })
    }

    override fun sayHello(request: Mono<Greeter.HelloRequest>): Mono<Greeter.HelloReply> {
        return request.flatMap { sayHello(it) }
    }
}
